crates/biome_js_analyze/src/lint/style/no_parameter_assign.rs

```diff
@@ -64,48 +116,142 @@ declare_lint_rule! {
+                    .map(ProblemType::ParameterReassignment)
+                    .collect();
+
+                signals.extend(param_reassignments);
             }
         }
-        // Empty iterator that conforms to `AllBindingWriteReferencesIter` type.
-        std::iter::successors(None, |_| None)
+
+        signals
     }
 
-    fn diagnostic(ctx: &RuleContext<Self>, reference: &Self::State) -> Option<RuleDiagnostic> {
-        let param = ctx.query();
-        Some(
-            RuleDiagnostic::new(
-                rule_category!(),
-                reference.syntax().text_trimmed_range(),
-                markup! {
-                    "Reassigning a "<Emphasis>"function parameter"</Emphasis>" is confusing."
-                },
-            )
-            .detail(
-                param.syntax().text_trimmed_range(),
-                markup! {
-                    "The "<Emphasis>"parameter"</Emphasis>" is declared here:"
-                },
-            )
-            .note(markup! {
-                "Use a local variable instead."
-            }),
-        )
+    fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        match state {
+            ProblemType::ParameterReassignment(expression) => Some(
+                RuleDiagnostic::new(
+                    rule_category!(),
+                    expression.syntax().text_trimmed_range(),
+                    markup! {
+                        "Reassigning a "<Emphasis>"function parameter"</Emphasis>" is confusing."
+                    },
+                )
+                    .detail(
+                        ctx.query().syntax().text_trimmed_range(),
+                        markup! {
+                        "The "<Emphasis>"parameter"</Emphasis>" is declared here:"
+                    },
+                    )
+                    .note(markup! {
+                    "Use a local variable instead."
+                }),
+            ),
+
+            ProblemType::PropertyAssignment(expression) => {
+                Some(
+                    RuleDiagnostic::new(
+                        rule_category!(),
+                        expression.syntax().text_trimmed_range(),
+                        markup! {
+                        "Assignment to a property of "<Emphasis>"function parameter"</Emphasis>" is confusing."
+                    })
+                        .note(markup! {"Use a local variable instead."}),
+                )
+            }
+        }
     }
 }
+
+fn extract_statement_from_reference(reference: &Reference) -> Option<JsExpressionStatement> {
+    reference
+        .syntax()
+        .ancestors()
+        .find_map(AnyJsStatement::cast)
+        .and_then(|stmt| match stmt {
+            AnyJsStatement::JsExpressionStatement(statement) => Some(statement),
+            _ => None,
+        })
+}
+
+#[derive(Debug)]
+pub enum ProblemType {
+    ParameterReassignment(Reference),
+    PropertyAssignment(AnyJsExpression),
```

Comment:
It seems that you only use their range n `diagnostic`.
We could thus just pass the text range:

```suggestion
    ParameterReassignment(TextRange),
    PropertyAssignment(TextRange),
```

Reply:
true